### ISSUE TITLE
r.now() and the non_atomic flag

### ISSUE THREAD

**Issue by nviennot** [rethinkdb/rethinkdb#6125]
The query: `r.table("table").get(1).update { { "expires_at" => r.now } }`

fails with: `Could not prove argument deterministic.  Maybe you want to use the non_atomic flag?`

I know the query may not be deterministic, but the update can still be performed in an atomic fashion. Can  the use of `r.now()` not require the use of the `non_atomic` flag?

**Comment by 1lann** [comment]
Although a somewhat hack, alternatively you can do
```js
r.now().do(function(now) {
	return r.table("table").get(1).update({"expires_at": now})
})
```

**Comment by nviennot** [comment]
That hack is not really an option here: https://github.com/nviennot/nobrainer/blob/master/lib/no_brainer/lock.rb#L124 :(

**Comment by chrisvariety** [comment]
I'm not sure the ruby syntax (are you just missing parenthesis?) but as far as I know `r.now()` doesn't require the use of the `non_atomic` flag. I just tried this query in the web UI data explorer and it works no problem:

`r.table('foobar').get('6d3c89fe-d5ab-4ccd-9db8-c10c680965da').update({createdAt: r.now()})`

**Comment by srh** [comment]
The difference with that query is that it computes `r.now()` before performing the update.  It becomes `something.update({createdAt: some_time_value})`.  Then all shards get updated with the same time value.

Correct me if my Ruby is wrong, but I think the query `r.table("table").get(1).update { { "expires_at" => r.now } }` passes a closure to update.  This closure gets transformed into a function in ReQL, and the query applies such a function to every document gets updated.  Thus, `r.now()` gets evaluated in the middle of the function on a document-by-document basis.  That makes it be non-deterministic.

The problem is that the update function needs to be run exactly once, so that all replicas get the same value.  That means the data gets retrieved from the shards, then the update function gets run, and then the result gets written back to the shards.  Because it has a round-trip across machines, this gets executed non-atomically.

A fix, if we want one, is for every query to have a pre-determined value that gets computed for `r.now()` in all uses in the query.  That might work for some queries, while being inappropriate for other uses such as change feeds.

**Comment by 1lann** [comment]
I should add that this issue only occurs in `next`, and not 2.3.5.

**Comment by nviennot** [comment]
The update can be done in an atomic fashion, regardless if the query is deterministic or not.

After all, `r.table("table").get(1).update { { "expires_at" => r.random } }` is not deterministic, and yet can still be performed in an atomic fashion.

**Comment by srh** [comment]
Hmm.  Well, running `r.random` atomically doesn't work for me in `v2.3.5` or `next`.

I think the cause of this bug is 47fe12a5fdeb27370b1079cb42b518e097b485f2.  See the changes to `src/rdb_protocol/time.cc`.  Previously, `r.now()` terms were turned into constant terms by replacing them in the AST (iirc). Now, they aren't.  The consequence is that `is_deterministic()`, called on such an expression, now produces a different result.

I think we can't change the `is_deterministic()` behavior, because `r.now()` terms *didn't* get replaced in secondary index functions, and we call `is_deterministic()` to determine what is a valid secondary index function.  I'll have to look.

**Comment by danielmewes** [comment]
Yeah it's a regression in `next`. `r.now` should work because it's replaced by a constant on the parsing node. It's like @srh says.

The change to make `r.now` no longer be a rewrite in the AST was to allow it to be included in the new write hook functions.
`r.now` should be deterministic, so we should overwrite `is_deterministic()` on it and make it return `true`. However we also decided that we wanted to disallow it in secondary index functions (and this is critical for correctness, because `r.now` determinism depends on the ReQL context which is not persisted for secondary index evaluation). I don't remember how exactly we thought we would ensure this. Maybe @nighelles remembers?
We will likely need to add another attribute to determine whether a term can be used in secondary indexes. It could default to forwarding `is_deterministic()`, but could be overwritten for `r.now()`.

**Comment by srh** [comment]
Fixed by #6396.
